File: Robot_In_Circle.py
"""
(PLEASE CORRECT ME IF MY TIME AND SPACE ARE WRONG)

Time Complexity : O(N) - length of instructions
Space Complexity : O(1)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
Your code here along with comments explaining your approach:
Here, we can define our directions as a dirs array in an anti clockwise manner. After iterating through all the instructions, if the
robot is stil facing north, it will never be able to come back to origin. Or else, if the robot reached origin , and is facing north,
then its ok.
"""

class Solution(object):
    def isRobotBounded(self, instructions):
        dirs = [(0, 1), (-1,0), (0, -1), (1, 0)]
        x, y = 0, 0
        d = 0
        
        for i in instructions:
            if i == "L":
                d = (d+1) % 4
            elif i == "R":
                d = (d + 3) % 4
            else:
                x += dirs[d][0]
                y += dirs[d][1]
        if d != 0 or ( x == 0 and y == 0):
            return True
        return False 

File: RobotinCircle.java
// Time Complexity : O(n)    
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// Approach

// We make use of dirs array
// When we encounter R or l we compute a value for i
// We add it to the x and y

class Solution {
    public boolean isRobotBounded(String instructions) {
        if(instructions == null || instructions.length() ==0)
            return true;
        
        int[][] dirs = {{0,1},{1,0},{0,-1}, {-1,0}};
        
        int x =0, y =0;
        int i =0;
        for(int j =0; j<instructions.length(); j++){
            char c = instructions.charAt(j);
            if(c == 'R')
                i = (i+1)%4;
            else if(c == 'L')
                i = (i + 3) % 4;
            else{
                x = x + dirs[i][0];
                y = y + dirs[i][1];
            }
        }
        return (x==0 && y==0) || i !=0;
    }
}